### PR TITLE
Feat/component poller suspend

### LINKED ISSUE #2301: Suspend update for PollingComponent
**Describe the problem you have/What new integration you would like**

When a device is playing music or voice and has a configured screen, there is a temporary pause in the sound playback when the screen refreshes. It appears as a quarter-second click. The following entries can be seen in the logs:
```
[12:06:00][V][component:204]: Component st7789v.display took a long time for an operation (0.37 s).
[12:06:00][V][component:205]: Components should block for at most 20-30ms.
```

**Please describe your use case for this integration and alternatives you've tried:**

I suggest adding a component with the ability to temporarily pause its operation (refresh). It doesn't necessarily have to be done for the screen; it would be beneficial to have this feature for all components. Here's an example of implementation:
```yaml
then:
  - component.suspend: main_display
  - delay: 2s
  - component.update: main_display
  - delay: 2s
  - component.resume: main_display
```
It is important to allow for manual updates of a component even when it is in a suspended state.

**Additional context**

Use voice_assistant: with display:

### PR DESCRIPTION
**Pull request by xdecock**
# What does this implement/fix?

This pr aims at allowing suspend / resume of polling component update, this is mainly in place to help out with "slow component" when managing audio interractions.

PS: Took a bit of time to understand how to start the version of my pull request, tested and it does work as i expected

Tested on a D1 Mini & and C3 Pico

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/feature-requests/issues/2301

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3205

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [x] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
esphome:
  name: test
  friendly_name: test
  platformio_options:
    board_build.flash_mode: dio
  on_boot:
    then:
      - logger.log: Wait
      - delay: 10s
      - component.suspend: adc_pin_3
      - logger.log: Stopped
      - delay: 10s
      - logger.log: Resume
      - component.resume: adc_pin_3
      - delay: 10s
      - logger.log: Stop 2
      - component.suspend: adc_pin_3
      - delay: 10s
      - logger.log: Start Slow
      - component.resume:
          id: adc_pin_3
          update_interval: 10s


esp32:
  board: lolin_c3_mini
  framework:
    type: esp-idf

# Enable logging
logger:
  level: VERBOSE
  hardware_uart: USB_SERIAL_JTAG

sensor:
  - platform: adc
    pin: GPIO3
    name: "Test 1"
    update_interval: 2s
    id: adc_pin_3

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by nagyrobi** [comment]
Would it be possible to start the poller in suspended state at boot?

The use case is, for example having to wait for a slave device to start up (as it's slower than the esp), before starting to poll it.

**Comment by xdecock** [comment]
> Would it be possible to start the poller in suspended state at boot?
> 
> The use case is, for example having to wait for a slave device to start up (as it's slower than the esp), before starting to poll it.

I thought about that, and imo, the most elegant way to do that, would be to add an action:
```yaml
  component.set_update_interval;
    id: component_id
    interval: xxxs
```

This mr do allow to "dynamically update the interval" 

The only thing i did not understand is how to create an action with that template.

That being, said, if i'm guided about how to implement those type of actions, i would probably do it as a separate pull request.

I'd then solve this exact scenario this way:
```
fast_component:
  id: fc_id
  update_interval: never
  
slow_component:
  id: sc_id
  on_startup:
    - component.set_interval:
      id: fc_id
      update_interval: 5s 
      resume: false #not sure about this or the next line
    - component.resume: fc_id
```

**Comment by jesserockz** [comment]
Instead of another action to change the interval, you could add an `Optional` config item to the `start` action for `update_interval`

**Comment by xdecock** [comment]
> Instead of another action to change the interval, you could add an `Optional` config item to the `start` action for `update_interval`

So more something in the vein of

```yaml
component.resume:
  id: fc_id
  update_interval: 3s
```

An extreme version would allow to merge those two actions via

```yaml
component.set_update_interval:
  id: fc_id
  update_interval: never

component.set_update_interval:
  id: fc_id
  update_interval: 5s

# possible helpers
component.set_update_interval:
  id: fc_id
  action: suspend

component.set_update_interval:
  id: fc_id
  action: resume
```

If this would work, i'd be happy with a pointer to the correct class to declare the correct structure in python, I'm quite happy to update my mr but i'm struggling a bit to find how things work.

**Comment by xdecock** [comment]
Fast tentative,

I'll debug it later today if it doesn't work, if the logic is sound, i'll redo my local tests, and update the documentation accordingly

**Comment by xdecock** [comment]
Updated the example, seems it works correctly on my local devices. i've struggled a bit to understand the code generation logic, so i'm happy with any correction i might have to do

File: esphome/automation.py [hunk 1]
```diff
@@ -303,6 +306,38 @@ async def component_update_action_to_code(config, action_id, template_arg, args)
...
+    return cg.new_Pvariable(action_id, template_arg, comp)
+
+
+@register_action(
+    "component.resume",
+    ResumeComponentAction,
+    maybe_simple_id(
+        {
+            cv.Required(CONF_ID): cv.use_id(cg.PollingComponent),
+            cv.Optional(CONF_UPDATE_INTERVAL): cv.positive_time_period_milliseconds,
```
Comment (jesserockz) [inline esphome/automation.py]: In the cpp class you define this as `templatable`, but the config side does not allow the use of a lambda to set.
```suggestion
            cv.Optional(CONF_UPDATE_INTERVAL): cv.templatable(cv.positive_time_period_milliseconds),
```

This also needs some changes in function to match

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/automation.py (hunk 1 above)
Comment (xdecock) [reply esphome/automation.py]: Updated, along side the test it runs as (i) expected it to work

**Comment by xdecock** [comment]
> Would it be possible to start the poller in suspended state at boot?
> 
> The use case is, for example having to wait for a slave device to start up (as it's slower than the esp), before starting to poll it.

this is now possible by starting with update_interval: never, and doing a component.resume when the slow component booted up, and set the interval with the resume action

**Review by jesserockz — APPROVED** [review]

**Pull request merged**